train.c

```diff
@@ -0,0 +1,94 @@
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include <float.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include "agents/temporal_difference.h"
+#include "game.h"
+#include "util.h"
+
+
+static double epsilon = 0.3;
+#define NUM_ITERATION 10000
+#define CALC_STATE_NUM(x)                 \
+    {                                     \
+        x = 1;                            \
+        for (int i = 0; i < N_GRIDS; i++) \
+            x *= 3;                       \
+    }
+
+#define RAND_UNIFORM ((float) rand() / (float) RAND_MAX)
+static unsigned int N_STATES = 1;
+
+static td_agent_t agent[2];
+
+int get_action_epsilon_greedy(char *table, td_agent_t *agent)
+{
+    int move_cnt = 0;
+    int *available_moves = get_available_moves(table, &move_cnt);
+    if (RAND_UNIFORM < epsilon) {  // explore
+        printf("explore %d\n", available_moves[rand() % move_cnt]);
+        return available_moves[rand() % move_cnt];
+    }
+    int act = get_action_exploit(table, agent);
+    epsilon *= GAMMA;
+    return act;
+}
+
+void train(int iter)
+{
+    char table[N_GRIDS];
+    memset(table, ' ', sizeof(char) * N_GRIDS);
+    int turn = (iter & 1) ? 0 : 1;  // 0 for 'O', 1 for 'X'
+    while (1) {
+        char win = check_win(table);
+        if (win == 'D') {
+            draw_board(table);
+            printf("It is a draw!\n");
+            break;
+        } else if (win != ' ') {
+            draw_board(table);
+            printf("%c won!\n", win);
+            break;
+        }
+        int cur_state_hash = hash(table);
+        int move = get_action_epsilon_greedy(table, &agent[turn]);
+        table[move] = "OX"[turn];
+        int nxt_state_hash = hash(table);
+        update_state_value(cur_state_hash, nxt_state_hash, &agent[0]);
+        update_state_value(cur_state_hash, nxt_state_hash, &agent[1]);
+        draw_board(table);
+        turn = !turn;
+    }
+}
+
+void store_state_value()
+{
+    FILE *fptr = NULL;
+    if ((fptr = fopen("state_value.bin", "wb")) == NULL)
```

Comment:
Consider using a macro to specify the file name instead of hard coding the file name directly in the code.